python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -238,14 +248,216 @@ private Node update(Node node) {
+ * `TPosOverflowNode` and `[**d]` is the synthesized dictionary argument `TKwOverflowNode`.
+ * There will be a store step from `2` to `[*t]` at pos `0` and one from `3` to `[**d]` at key
+ * `a`.
+ *
+ * For the call
+ * ```python
+ * f(0, **{"y": 1, "a": 3})
+ * ```
+ * no tuple arguemnt is synthesized. It is modelled as
+ * ```python
+ * f(0, [y=1], [**d])
+ * ```
+ * where `[y=1]` is the synthesised unpacked argument `TKwUnpacked` (with `name` = `y`). There is
+ * a read step from `**{"y": 1, "a": 3}` to `[y=1]` at key `y` to get the value passed to the parameter
+ * `y`. There is a dataflow step from `**{"y": 1, "a": 3}` to `[**d]` to transfer the content and
+ * a clearing of content at key `y` for node `[**d]`, since that value has been unpacked.
+ */
+module ArgumentPassing {
+  /**
+   * Holds if `call` represents a `DataFlowCall` to a `DataFlowCallable` represented by `callable`.
+   *
+   * It _may not_ be the case that `call = callable.getACall()`, i.e. if `call` represents a `ClassCall`.
+   *
+   * Used to limit the size of predicates.
+   */
+  predicate connects(CallNode call, CallableValue callable) {
+    exists(DataFlowCall c |
+      call = c.getNode() and
+      callable = c.getCallable().getCallableValue()
+    )
+  }
+
+  /**
+   * Gets the `n`th parameter of `callable`.
+   * If the callable has a starred parameter, say `*tuple`, that is matched with `n=-1`.
+   * If the callable has a doubly starred parameter, say `**dict`, that is matched with `n=-2`.
+   * Note that, unlike other languages, we do _not_ use -1 for the position of `self` in Python,
+   * as it is an explicit parameter at position 0.
+   */
+  NameNode getParameter(CallableValue callable, int n) {
+    // positional parameter
+    result = callable.getParameter(n)
+    or
+    // starred parameter, `*tuple`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -1 and
+      result = f.getVararg().getAFlowNode()
+    )
+    or
+    // doubly starred parameter, `**dict`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -2 and
+      result = f.getKwarg().getAFlowNode()
+    )
+  }
+
+  /**
+   * Gets the node representing the argument to `call` that is passed to the parameter at
+   * (zero-based) index `paramN` in `callable`. If this is a positional argument, it must appear
+   * at index `argN` in `call`.
+   *
+   * `argN` will differ from `paramN` for method- or constructor calls, where the first parameter
+   * is `self` and the first positional argument is passed to the second positional parameter.
+   * Similarly for classmethod calls, where the first parameter is `cls`.
+   *
+   * NOT SUPPORTED: Keyword-only parameters.
+   */
+  Node getArg(CallNode call, int argN, CallableValue callable, int paramN) {
+    connects(call, callable) and
+    paramN - argN in [0, 1] and // constrain for now to limit the size of the predicate; we only use it to insert one argument (self).
```

Comment:
This line is giving me a real :grimacing: vibe. For one, it seems to invalidate the specification set out in the QLDoc comment: for `foo.bar(1,2,3)` we have two values of `paramN` for each positional argument, but (depending on whether this is a method call or not) only one of these values is valid. Thus, it's not the case that "````If this is a positional argument, it must appear at index `argN` in `call`.````". (In particular, "`must`" is too strong.)

I think there must be a better way of doing this. At the very least, it seems like it should be possible to factor out the `[0,1]` bit in its own argument as an "argument index correction", and then do away with either `paramN` or `argN`.

Size-wise this amounts to the same thing (and is still a bit icky), but perhaps it would be less confusing?

Reply:
I was a bit worried about this part of the predicate as well until I realized that the predicate would only be used by other classes defined inside `DataFlowPrivate`, where they always make it obvious if there is a +1 or not. I guess marking this predicate as `private` to highlight that it really just is a private helper predicate would be a good idea?

Reply:
During a meeting with `@tausbn` we came up with a solution involving binding sets, which seems good for now; at least it eliminates the offending line. I also marked the whole `ArgumentPassing` module as `private`.

Reply:
Went with the IPA type solution